tempo tracing interface spec (#41)
canonical/charm-relation-interfaces: interfaces/tracing/v0/README.md

# `tracing`

## Usage

This relation interface describes the expected behavior of any charm claiming to be able to provide or consume a tempo-compliant tracing server.

## Direction

Tracing is done in a push-based fashion. 
The receiving endpoint of the tracing backend, also referred to as an ingester, can support a number of different protocols, such as [otlp-grpc](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlpgrpc) and [otlp-http](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).
The tracing backend exposes, for each protocol it supports, an endpoint at which the server is ready to accept that protocol. 

> So the directionality of the relation flows from the observed, the application producing the traces, to the observer (aka the trace ingester, tha is, a Tempo(-compliant) backend).

Reviewer: ```suggestion
So the directionality of the relation flows from the observed, the application producing the traces, to the observer (aka the trace ingester, that is, a Tempo(-compliant) backend).
```

# `tracing`

## Usage

This relation interface describes the expected behavior of any charm claiming to be able to provide or consume a tempo-compliant tracing server.

## Direction

Tracing is done in a push-based fashion. The receiving endpoint of the tracing backend, also referred to as an ingester, can support a number of different protocols, such as [otlp-grpc](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlpgrpc) and [otlp-http](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).

> The tracing backend **provides**, for each protocol it supports, an endpoint at which the server is ready to accept that protocol. So the directionality of the relation flows from the observer: the trace ingester (a Tempo(-compliant) backend), to the observed: the application producing the traces, such as a mattermost charm.

Reviewer 2: Strictly speaking, a relation does not have directionality.

Author: fair enough

This follows the convention of other push models such as `loki_push_api`.

We call the data structure that is exchanged via this interface 'tracing backend'.

```mermaid
flowchart LR

>     Provider -- TracingBackend --> Requirer

Reviewer 2: Seems clearer when we use the domain's terms next to provider/requirer:
```suggestion
    provider[Ingester (provider)] -- TracingBackend --> requirer[Observed app (requirer)]
```

```

## Behavior
### Provider

> - Is expected to publish the hostname at which the server is reachable.

Reviewer 2: "Hostname" seems to restricting as it excludes scheme, path, ports etc.
"Address" is too ambiguous.
How about URL?

```suggestion
- Is expected to publish the URL at which the server (ingester) is reachable.
```

- Is expected to run a server supporting one or more tracing protocols such as [OTLP](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#opentelemetry-protocol-specification).
- Is expected to publish, for each protocol it accepts, the port at which the server is listening.

### Requirer

> - Is expected to push traces to one or more of the supported endpoints.

Reviewer 2: "push traces" is a workload capacity, not a relation/charm capacity.
Iiuc, in the context of this repo, "requirer" and "provider" do not include the workload.

Author: in my experience this distinction is kind of blurry in this repository, although it does make sense to me.

- Is expected to handle cases where none of the protocols offered by the provider is supported. 

## Relation Data
### Provider

The provider exposes via its application databag a single hostname, at which the server is reachable, and a list of `ingester-ports`.
Each ingester port supports a certain tracing protocol, such as OTLP_GRPC or Jaeger. 
The full list of supported trace protocols can change with time, but those supported by Tempo at the time of writing are:

- tempo
- otlp_grpc
- otlp_http
- jaeger
- zipkin

[\[JSON Schema\]](./schemas/provider.json)

#### Example
```yaml
# unit_data: <empty> 
application_data: 

>   hostname: "http://foo.bar/my-model-my-unit-0"

Reviewer 2: Strictly speaking, "hostname" would be just `foo.bar`.

Need to figure out how to specify [URL parts](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse) in relation data.
If the port is different, is there guarantee that the other URL parts remain the same?

  ingester-ports: 
    - type: otlp_grpc
      port: 1234
    - type: otlp_http
      port: 5678
```

### Requirer

> The requirer side is not expected to publish any data via this relation's databags.

Reviewer 2: Haven't looked into tracing yet, but would we want the requirer to announce its juju topology for labelling of traces?

Author: The thing producing the traces would be responsible for injecting that kind of metadata in the traces. So no relabeling is required.

# `tracing`

## Usage

> This relation interface is meant to exchange the necessary information for a charm to integrate with a Tempo-compliant tracing backend.

Reviewer 3: It's not just for the charm that integrates with a tempo-compliant tracing backend, it's also for that tempo-compliant tracing backend. I think the generic `This relation interface describes the expected behavior of any charm claiming to be able to provide or consume a tempo-compliant tracing server.`

Reviewer: Agreed.

> This relation interface is meant to exchange the necessary information for a charm to integrate with a Tempo-compliant tracing backend.

Author: ```suggestion
This relation interface describes the expected behavior of any charm claiming to be able to provide or consume a tempo-compliant tracing server.
```

## Direction

Tracing is a push-based system. The tracing backend, also called an ingester, can support a number of different protocols, such as [otlp-grpc](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlpgrpc) or [otlp-http](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).
The tracing backend **provides**, for each protocol it supports, an endpoint at which the server is ready to accept that protocol. So the directionality of the relation flows from the observer: the trace ingester (a Tempo(-compliant) backend), to the observed: the application producing the traces, such as a mattermost charm.

> The requirer is supposed to receive the supported protocols and use it to determine where to send the traces.

Reviewer: ```suggestion
```

Already described below in the behavior of the requirer.

This follows the convention of other push models such as `loki_push_api`.

We call the data structure that is exchanged via this interface 'tracing backend'.

```mermaid
flowchart LR
    Provider -- TracingBackend --> Requirer
```

## Behavior
### Provider

- Is expected to publish the hostname at which the server is reachable.
- Is expected to run a server supporting one or more tracing protocols such as [OTLP](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#opentelemetry-protocol-specification).

> - Is expected to publish via relation data, for each protocol it accepts, the port at which the server is listening.

Reviewer: ```suggestion
- Is expected to publish, for each protocol it accepts, the port at which the server is listening.
```
